File: frontend/src/types/graph.ts
export type GraphNodeType = "repository" | "file" | "class" | "method";

export interface GraphNode {
  id: string;
  type: GraphNodeType;
  label: string;
  metadata: Record<string, string | number | null | undefined>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
}

export interface GraphStats {
  files_count: number;
  classes_count: number;
  methods_count: number;
  functions_count: number;
}

export interface RepositoryGraph {
  graph_type: string;
  branch: string | null;
  nodes: GraphNode[];
  edges: GraphEdge[];
  stats: GraphStats;
  empty_state: string | null;
}

export interface BlastRadiusNode {
  file_path: string;
  hop: number;
}

export interface BlastRadiusResponse {
  file_path: string;
  direction: "dependents" | "dependencies";
  max_depth: number;
  branch: string | null;
  nodes: BlastRadiusNode[];
  total: number;
}

export interface GraphPathResponse {
  source_file: string;
  target_file: string;
  max_depth: number;
  branch: string | null;
  paths: string[][];
  total_paths: number;
}

File: backend/app/schemas/graph.py
from typing import Any, Literal

from pydantic import BaseModel, Field

GraphNodeType = Literal["repository", "file", "class", "method"]


class GraphNode(BaseModel):
    id: str
    type: GraphNodeType
    label: str
    metadata: dict[str, Any]


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str


class GraphStats(BaseModel):
    files_count: int
    classes_count: int
    methods_count: int
    functions_count: int = 0


class RepositoryGraphResponse(BaseModel):
    graph_type: str = "structure"
    branch: str | None = None
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    stats: GraphStats
    empty_state: str | None = None


class BlastRadiusNode(BaseModel):
    file_path: str
    hop: int


class BlastRadiusResponse(BaseModel):
    file_path: str
    direction: str
    max_depth: int
    branch: str | None = None
    nodes: list[BlastRadiusNode] = Field(default_factory=list)
    total: int = 0
    message: str | None = None
    suggested_direction: str | None = None


class GraphPathResponse(BaseModel):
    source_file: str
    target_file: str
    max_depth: int
    branch: str | None = None
    paths: list[list[str]] = Field(default_factory=list)
    total_paths: int = 0
    bidirectional: bool = False
    message: str | None = None


class ImportEdgeItem(BaseModel):
    source_path: str
    target_path: str
    dependency_type: str


class ImportGraphSummaryResponse(BaseModel):
    branch: str | None = None
    edges: list[ImportEdgeItem] = Field(default_factory=list)
    connected_files: list[str] = Field(default_factory=list)
    source_files: list[str] = Field(default_factory=list)
    target_files: list[str] = Field(default_factory=list)
    total_edges: int = 0
